In [3]:
import torch
from torch import nn
from torch.nn import functional as F

# [`conv1d`](http://pytorch.org/docs/stable/nn.html#torch.nn.functional.conv1d)

* `inputs`: (`batch_size`, `in_channels`, `data_length`)
* `filters`: (`out_channels`, `in_channels`, `filter_length`)

Default: 

* No padding
* Stride 1

In [7]:
inputs = torch.randn(20, 16, 50)
filters = torch.randn(33, 16, 3)
output = F.conv1d(inputs, filters)

In [6]:
print(output.shape)

torch.Size([20, 33, 48])


Description: 

1. The filters are of length 3. We want 33 output channels, and we currently have 16 input channels.
2. The first output channel is created by sliding one filter over each input channel and summing them up.
3. This is then done 32 more times. Thus, we end up with 33 output channels.
4. We can repeat this process for each image in our input batch.

# [`conv2d`](http://pytorch.org/docs/stable/nn.html#torch.nn.Conv2d)

* `inputs`: (`batch_size`, `in_channels`, `size_h`, `size_w`)
* `filters`: (`out_channels`, `in_channels`, `size_h`, `size_w`)

Default: 

* No padding
* Stride 1

In [9]:
inputs = torch.randn(1,4,5,5)
filters = torch.randn(8,4,3,3)
output = F.conv2d(inputs, filters, padding=1)

In [10]:
print(output.shape)

torch.Size([1, 8, 5, 5])


Description: 

1. The filters are of size 3x3. We want 8 output channels, and we currently have 4 input channels.
2. The first output channel is created by sliding one filter over each of the 4 input channels and summing them up.
3. This is then done 7 more times. Thus, we end up with 8 output channels.
4. We can repeat this process for each image in our input batch.

Rules of thumb on image size:

* With filter size $n$, $n$ odd, if we use padding $\lfloor \frac{n}{2} \rfloor$ and stride 1, then we get the same output size as input size.
* With filter size $n$, $n$ even, if we use padding $\frac{n}{2}$ and stride 1, then we get the same output size as input size.

# [`conv2dtranspose`](http://pytorch.org/docs/stable/nn.html#torch.nn.ConvTranspose2d)

* `inputs`: (`batch_size`, `in_channels`, `size_h`, `size_w`)
* `filters`: (`out_channels`, `in_channels`, `size_h`, `size_w`)

Default: 

* No padding
* Stride 1

In [15]:
# With square kernels and equal stride
m1 = nn.ConvTranspose2d(16, 33, 3, stride=2)
input1 = torch.randn(20, 16, 50, 100)
output1 = m1(input)

In [16]:
print(output1.shape)

torch.Size([20, 33, 101, 201])


Description: 

1. We want an input size that will have us end up with batch size 20, with 16 channels, height 50, and width 100, if we apply to it a convolution operation 16 output channels, 33 input channels, filter size 3 and stride 2.
2. It turns out to do this, we can start with batch size 20, 33 channels (both obvious), and, because of the geometry of the convolution, height 101 and width 201.

**Rules of thumb**:

Similar rules of thumb apply: if we want to add deconvolution layers to a network to make it deeper, adding `ConvTranspose2d` operations with filter size $n$ and padding $\lfloor \frac{n}{2} \rfloor$ can make your network deeper without messing with your image size.

In [17]:
input2 = torch.randn(1, 16, 28, 28)
m2 = nn.ConvTranspose2d(16, 32, 3, padding=1)
output2 = m2(input2)
print(output2.shape)

torch.Size([1, 32, 28, 28])
